### ISSUE TITLE
[CuDNN Attention] Performance Grouped Query Attention

### ISSUE THREAD

**Issue by drisspg** [pytorch/pytorch#139586]
# Summary
We recently landed support for grouped query attention via use `enable_gqa` on sdpa, however this is only enabled on the flash attention backend.  This leads to a weird situation where it could have been more beneficial for a user to have not used the enable GQA flag and called repeat interleave prior to calling SDPA in order to use the CUDNN net backend.  

It looks  Like there is explicit support for the GQA situation in the CUDI and NAPI, we should add support for this. 
https://docs.nvidia.com/deeplearning/cudnn/latest/api/cudnn-graph-library.html#cudnn-backend-operation-reduction-descriptor

cc @msaroufim @mikaylagawarecki @jainapurva, @eqy, @Skylion007

**Comment by eqy** [comment]
Do you have some code samples for exercising this properly? I wasn't quite sure what the difference was between `enable_gqa=True/False` when heads_k/heads_v < heads_q.

**Comment by eqy** [comment]
Right now the behavior of the cuDNN backend is `enable_gqa` seems to do nothing, as we do not reject nheads being different as long as they are valid factors
```
import torch

import torch
from torch.nn.functional import scaled_dot_product_attention
from torch.nn.attention import SDPBackend, sdpa_kernel

batch = 4
seq_len_q = 512
seq_len_kv = 1024
D = 128
# Sample call to SDPA - GQ
query = torch.rand(batch, 32, seq_len_q, D, device='cuda', dtype=torch.bfloat16)
key = torch.rand(batch, 8, seq_len_kv, D, device='cuda', dtype=torch.bfloat16)
value = torch.rand(batch, 8, seq_len_kv, D, device='cuda', dtype=torch.bfloat16)
with sdpa_kernel([SDPBackend.MATH]):
   output_math = scaled_dot_product_attention(query, key, value, is_causal=True, enable_gqa=True)

with sdpa_kernel([SDPBackend.CUDNN_ATTENTION]):
   output_cudnn = scaled_dot_product_attention(query, key, value, is_causal=True, enable_gqa=True)

torch.testing.assert_close(output_math, output_cudnn)

#with sdpa_kernel([SDPBackend.MATH]):
#   output_math2 = scaled_dot_product_attention(query, key, value, is_causal=True, enable_gqa=False)

with sdpa_kernel([SDPBackend.CUDNN_ATTENTION]):
   output_cudnn2 = scaled_dot_product_attention(query, key, value, is_causal=True, enable_gqa=False)

torch.testing.assert_close(output_math, output_cudnn2)
```
So for `MATH` is `enable_gqa` just enabling another set of shape checks? Another way to phrase that is there any ambiguity when nheads q != nheads k / nheads v for other behavior or is it OK to assume the user will only mean one possible thing like what is done for grouped/depthwise convolution?

I also wasn't able to run the above with `FLASH_ATTENTION` or `EFFICIENT_ATTENTION` specified in the context manager, is that supposed to be done with another method?

**Comment by Skylion007** [comment]
> This leads to a weird situation where it could have been more beneficial for a user to have not used the enable GQA flag and called repeat interleave prior to calling SDPA in order to use the CUDNN net backend.

Theoretically possible with a custom random_mask to get around the fact FA3 kernals do not have dropout support yet for when we implement that too

**Comment by drisspg** [comment]
> Right now the behavior of the cuDNN backend is enable_gqa seems to make no difference, as we do not reject nheads being different as long as they are valid factors

This was kinda a reminder issue for me to confirm, in theory we should reject the non valid factors unless enable_gqa =True. I think in general it is pretty clear with the default B,H,S,D interpretation. However, this is not valid broadcasting by default and it was not until we made the enable_gqa change that added the explicit meaning of 'H' to dim[-3]

**Comment by mseeger** [comment]
Did you guys realize that the `enable_gqa=True` feature is required to make the recently very popular DeepSeek models reach their potential?

For example, their model uses 128 heads, but the key and value tensors do not depend on the heads (call this GQA if you like). Not exploiting this means you require 128x more memory on GPU than is needed.

I'd also suspect one can make use of this in a low-level CUDA implementation to improve computational speed.

I'll contribute efficient multi-head latent attention to Hugging Face, but this will not really work well if `enable_gqa=True` is not widely supported. I hope this provides some motivation: It is not just GQA.

**Comment by drisspg** [comment]
We have fixed this
